### DISCUSSION TITLE
Tidal Connect installed on Moode Audio

### DISCUSSION THREAD

**Discussion by GioF71** [GioF71/tidal-connect#5] [Announcements]
I just documented the process of installing Tidal Connect on a MoodeAudio box with a Hifiberry HAT.  
Everything should be available in the README.md.
I also discovered and fixed a bug when using card index in the process, so that has been fun.
Enjoy

**Comment by DKrampi** [comment]
Hi!
I need help :) 
After i have run the following script 
sudo apt update
sudo apt install docker.io docker-compose
sudo usermod -a -G docker $USER

and after this script

pi@moode:/ $ ./configure.sh -n "AUDIO" -f "Tidal" -m "Moode"
-bash: ./configure.sh: No such file or directory     :(

from which directory should I run the configure.sh script? or what else would I need to do to configure the default sound device?
Thank you! Dí

**Comment by GioF71** [comment]
Hello, you need to clone the repository first.
Execute the following:

```
sudo apt install -y git
cd
git clone https://github.com/GioF71/tidal-connect.git
cd tidal-connect
./configure.sh -n "AUDIO" -f "Tidal" -m "Moode"

```

Let me know if this helps
Cheers

**Reply by DKrampi** [reply]
the script seems to run fine :)
Thank you! What comes next?
Unfortunately, I don't have a Tidal account yet but I'll get one soon ;)

**Reply by GioF71** [reply]
You need the tidal account in order to play to Tidal Connect. Why not trying with a free trial?

**Comment by GioF71** [comment]
BTW why "AUDIO" for the "-n" parameter? Is that the name of the sound card?
It might help if you post the contents of

```
cat /proc/asound/cards
```

**Reply by DKrampi** [reply]
thats why :)
pi@moode:~/tidal-connect $ cat /proc/asound/cards
 0 [b1             ]: bcm2835_hdmi - bcm2835 HDMI 1
                      bcm2835 HDMI 1
 1 [Headphones     ]: bcm2835_headpho - bcm2835 Headphones
                      bcm2835 Headphones
 2 [AUDIO          ]: USB-Audio - SMSL USB AUDIO
                      SMSL SMSL USB AUDIO at usb-0000:01:00.0-1.2, high speed
pi@moode:~/tidal-connect $

**Reply by GioF71** [reply]
I don't know why you marked the text as deleted but AUDIO as the audio cards seems fine to me. Does this work?

**Reply by DKrampi** [reply]
Do I need to run anything else to make it work? Should I have to run "docker-compose up -d"? 
I'm waiting to get my friend's Tidal account details, until then I can't try it. The Moode works, of course. :)

**Reply by GioF71** [reply]
Nothing else. You don't need credentials on the tidal-connect device. The app on the device is authenticated and will be able to use tidal-connect devices.
Remember if you choose a tidal connect instance as the output on a Tidal app, the audio card will be locked exclusively. You will need to disconnect the tidal app in order to use the other functionalities of Moode.

**Reply by DKrampi** [reply]
In the meantime, just in case, I ran the docker-compose up -d command, it ran fine.
After on/power off, does Mood-al automatically start the tidal-connect or do I need to start it via ssh?

**Reply by GioF71** [reply]
yes, it will restart, because on the docker-compose file there is a line saying that container restart mode is set to "unless-stopped".
So it will run unless you stop manually with a 

docker-compose stop

or 

docker-compose down

**Reply by DKrampi** [reply]
I downloaded the Tidal app for PC, signed in. Unfortunately the Sound Output does not show Moode, only the PC System default

**Reply by GioF71** [reply]
Quite strange, are you on the same network? If you are running windows, is your network "private"? Setting it to public might affect discoverability.
Maybe try installing Tidal on a android phone, an iphone of a tablet if you can.
Also corporate PCs can have issues (mine also does not discover Tidal Connect)

**Reply by GioF71** [reply]
Also by the strings you used with configure.sh, your device should be named "Tidal"

**Reply by DKrampi** [reply]
I downloaded the Tidal app to my mobile phone, I'm on the same network/router/Wi-Fi with Moode, only This Device is shown, no other device :(

**Reply by DKrampi** [reply]
Is there anything to adjust on Moode?

**Reply by GioF71** [reply]
Can you post the output of:

docker-compose logs

from the tidal-connect directory

**Reply by GioF71** [reply]
There seems to be a space after AUDIO in the output of cat /proc/asound/cards, can you verify that?

**Reply by DKrampi** [reply]
there was no space

**Comment by GioF71** [comment]
You may also want to have a look to the upmpdcli docker image (https://github.com/GioF71/upmpdcli-docker), there is a new Tidal plugin which will allow you to create a media server dedicated to Tidal. See discussion [here](https://github.com/GioF71/upmpdcli-docker/discussions/281)

**Reply by DKrampi** [reply]
I also installed this, upmpdcli.  How to manage, configure?

**Reply by GioF71** [reply]
About upmpdcli:

Just create a directory, maybe on moode itself, named e.g. "upmpdcli"
Create a file named "docker-compose.yml" with the contents from the discussion.
Create the credentials using the following command:

docker run --rm -it --entrypoint /app/bin/get-tidal-credentials.py giof71/upmpdcli

Open link that appears on screen, authenticate and authorize on the browser.
Put the contents that will displayed in a '.env' file in the same directory.

Try a 

docker-compose up -d

**Reply by DKrampi** [reply]
I restarted Moode and now discovered Tidal on my mobile phone and it sounds, in MQA too! :)
A thousand thanks and thanks!!!

**Reply by GioF71** [reply]
Good to hear. What was the problem? The space in the device name?
If you can, please consider starring the repo!

**Reply by DKrampi** [reply]
I really don't know what happened. :) what was the reason for the improvement. 
Apart from running this "docker pull giof71/upmpdcli" and restarting Moode, I did nothing.

**Reply by DKrampi** [reply]
"If you can, please consider starring the repo!" I do not understand, you mean promote your tidal-connect solution? please explain what you want. I am happy to promote it on the forums I know.
Sorry, I do not understand the language of today :) And English is not my strong point either.
I cannot express my gratitude enough.

**Reply by GioF71** [reply]
Hello, thank you for the kind words.
Starring means click the small star icon on the top-right corner of the repository page (in this case [this](https://github.com/GioF71/tidal-connect), where it says "Star" with the current number of stars received. When you star it, the icon becomes yellow, yes of course it is for appreciation.

**Reply by GioF71** [reply]
Ok, so maybe you probably just did not restart the container after changing the configuration, at least I am guessing that this a possibility!
Maybe I need to clarify this in the documentation.

**Reply by DKrampi** [reply]
I think it would be useful to describe the whole process (the whole command line from start to finish) in the documentation, so that people who are less familiar with the process - like me - can successfully install tidal-connect :)

**Comment by GioF71** [comment]
> I think it would be useful to describe the whole process (the whole command line from start to finish) in the documentation, so that people who are less familiar with the process - like me - can successfully install tidal-connect :)

First of all, thank you for starring the repo.
About the suggestion, you are right, I have just updated the documentation. Can you take a look and tell me if any of the steps require further documentation?

Cheers

**Reply by DKrampi** [reply]
Hi, 
I see what looks like a missing 'cd tidal-connect' command in front of
./configure.sh -n "AUDIO" -f "Tidal" -m "Moode"
Configure Audio
,
,
Great, the operating system has just disabled the onboard audio and set the Hifiberry HAT as the default card.
So let's configure Tidal Connect:
.....

As soon as I have time, I'll try to see if the updated description works :)

**Comment by GioF71** [comment]
Ok, I have added "cd $HOME/tidal-connect" where needed (hopefully), let me know!

**Reply by DKrampi** [reply]
I re-did a tidal-connect installation following the updated instructions, successfully. :)

**Reply by GioF71** [reply]
Good! Feedbacks like yours are quite useful for improving the documentation. Thank you.

**Comment by longday11** [comment]
Hi - unfortunately, since a couple of days ago, this solution has stopped working for me. 
9 out of 10 tracks just sit there at zero and won't play. Eventually one will. So bizarre.
Has anyone else experienced this? Such a shame!
To fix, I've tried to do a fresh reinstall of the latest moode (9.1.4) on my pi4, and have re-installed this solution - but same result. Tidal app on phone connects to Tidal Connect but tracks just will not play. I didn't really want to splash out for a wiim - I like to tinker.

**Comment by bauhorm** [comment]
Hello,  I'm DIYer from Thailand . I've found out issue that My DAC only display 48KHz  maximum bit rate only.  and additional question if tidal connect can pass http stream to MPD player instead to go direct to alsa?

**Comment by GioF71** [comment]
Hello, unfortunately this solution is quite limited, see [here](https://github.com/GioF71/tidal-connect#mqa-content-is-gone).
There are alternatives though, some involve mpd, see [here](https://github.com/GioF71/tidal-connect#alternatives).

About mpd, what you ask is not possible AFAIK. Even if, maybe captured from the logs, a stream might be playable by mpd (but I never tried), the application would not be able to control mpd.

I have built a Tidal plugin for upmpdcli, maybe give it a try (sample configuration [here](https://github.com/GioF71/audio-tools/tree/main/media-servers/tidal-hires)).

Cheers

**Comment by kabexa22** [comment]
Hi, thanks for this. I really had to struggle, but I got it working. The key is to get the DAC name right. I also happen to have iqaudio, and there's a sample file. The one I had to configure to get it working is fine. If anyone finds it useful, the ASOUND_FILE_PREFIX parameter is helpful. Thanks!

**Comment by ponim** [comment]
Hello! lately I realized that my Moode with the lovely Tidal docker designed by you, stopped playing in Hi-Res, even 16bit/44100kBps. it can be easily seen in Windows Tidal application, where the rate window shows LOW instead of HIGH. I opened the ticket to Tidal support, but they still did not figured out the issue. 

Can anything be done on the docker side to fix that?

**Comment by GioF71** [comment]
Hello @ponim, I just tried, an 16/44 works on my side. Can you post the container logs?
BTW it's not my solution, I have only built an alternative entrypoint script to make it easier to run the existing docker image with (almost) any alsa device.
Please, don't open issues on Tidal official boards about this solution. It is completely unsupported. Maybe investigate about its origin, and you will understand why Tidal will be a lot happier when they are able to phase it out :-)

**Comment by mike-golant** [comment]
hi 
I have had the same issue since last week( or a week + a half :) )
Which kind of logs do you need?

**Comment by mike-golant** [comment]
attaching Docker-Compose logs 
[logs.txt](https://github.com/user-attachments/files/26356215/logs.txt)
I have a DIY Hi-Fi DAC connected to RPI4  via I2S

**Comment by GioF71** [comment]
this line is scary:

```text
[2Ktidal-connect  | [2026-03-30 12:35:19.862] [tisoc] [warning] [PlaybackEngineTidal.cpp:142] url fetch error: 2, 401, {"status":401,"subStatus":11003,"userMessage":"The token has expired. (Expired on time)"}
```

Can you try to remove CARD_FORMAT=S24_LE from your .env file? usually we can use S16_LE or S32_LE. You might as well not need the parameter at all. But S24_LE is problematic AFAIK, some DACs don't work with that setting.
Restart using

```text
docker-compose up -d --force-recreate
```

A re-create always happens if you change the value of any env variable. The --force-recreate tell docker to always create the container from scratch.

**Comment by mike-golant** [comment]
[logs2.txt](https://github.com/user-attachments/files/26357420/logs2.txt)
[logs3.txt](https://github.com/user-attachments/files/26357421/logs3.txt)

Hi again :)
new logs :
logs2.txt with CARD_FORMAT=S32_LE
logs3.txt with CARD_FORMAT= 

same issue ...

**Reply by ponim** [reply]
and I got the same issue when setting is S16_LE

**Reply by GioF71** [reply]
post your logs as well if you can, thank you

**Comment by GioF71** [comment]
at least that "certificate" error is gone. any particular track you're trying to play?

**Reply by mike-golant** [reply]
![image](https://github.com/user-attachments/assets/acabeeeb-2190-48b1-8cfe-12aaf21ed9cf)

Attaching screenshots

**Reply by ponim** [reply]
<img width="1792" height="953" alt="Screenshot 2026-03-24 at 15 15 33" src="https://github.com/user-attachments/assets/62bf3b03-650e-4474-8425-7a5589eac210" />

in the picture you can observe how we see the issue (also hear :) ). 
[ponim_log.txt](https://github.com/user-attachments/files/26357956/ponim_log.txt)

**Reply by ponim** [reply]
as I told, we see the issue as LOW indicator in the Tidal application's right bottom corner. When using the phone, instead of usual HIGH indicator, no indication at all (as on the picture Mike sent). Can we check the real stream rate in the command line?

**Comment by GioF71** [comment]
@ponim you seem to be using one of my sample config files, which specifies CARD_NAME=DAC along with `FRIENDLY_NAME=Aune S6 USB DAC`.

See in the log:

```text
tidal-connect  | CARD_NAME=DAC
```

But also from your logs you seem to have some hifiberry dac:

```text
tidal-connect  | Skipping audio device [sndrpihifiberry] at index [0]
```

maybe write the following line in your `.env` file:

```text
CARD_NAME=sndrpihifiberry
```

Also a multi-word friendly name has represented an issue in the past, try using just a word (or just no spaces).

**Reply by ponim** [reply]
right - I just did the fresh re-install of the docker today. Unfortunately, it did too work (just 360 kbps) even with the correct parameters. I need to mention - that the issue appeared just 1-2 weeks ago, **with no any changes from my side in the Moode or docker settings...** I just put my attention on the LOW indicator in the application and the sound that like mp3...

**Reply by GioF71** [reply]
> right - I just did the fresh re-install of the docker today. Unfortunately, it did too work (just 360 kbps) even with the correct parameters. I need to mention - that the issue appeared just 1-2 weeks ago, **with no any changes from my side in the Moode or docker settings...** I just put my attention on the LOW indicator in the application and the sound that like mp3...

Ok, understood. It's just not happening here on my device (also moode, with hifiberry digi+). So weird.

**Reply by ponim** [reply]
what do you see - HIGH, 96kHz? can you share the screen shot please?

**Reply by GioF71** [reply]
> what do you see - HIGH, 96kHz? can you share the screen shot please?

I just see "high". See:

![Screenshot_20260330_220653_com_aspiro_tidal_MainActivity](https://github.com/user-attachments/assets/5f80cbda-3d1c-4083-ae5e-262e1b6741bb)

**Reply by ponim** [reply]
oh, really strange. do you use LAN or Wi-Fi?

**Reply by GioF71** [reply]
I use wired network

**Reply by ponim** [reply]
<img width="1179" height="2556" alt="image" src="https://github.com/user-attachments/assets/0192a4a3-7351-44c5-9b57-60983b9703d1" />
This is mine…

**Reply by ponim** [reply]
can I ask you to share your docker-compose.yaml file please?

**Reply by GioF71** [reply]
> can I ask you to share your docker-compose.yaml file please?

sure, it's the current master, unmodified. Also following my .env.

```text
pi@moode-living-digi-desktop:~/git/tidal-connect $ cat docker-compose.yaml 
---
version: "3"

services:
  tidal-connect:
    image: ${TIDAL_CONNECT_IMAGE:-edgecrush3r/tidal-connect:latest}
    container_name: tidal-connect
    network_mode: host
    devices:
     - /dev/snd
    environment:
     - CARD_NAME=${CARD_NAME:-}
     - CARD_INDEX=${CARD_INDEX:-}
     - CARD_DEVICE=${CARD_DEVICE:-}
     - CARD_FORMAT=${CARD_FORMAT:-}
     - CERTIFICATE_PATH=${CERTIFICATE_PATH:-}
     - CLIENT_ID=${CLIENT_ID:-}
     - CREATED_ASOUND_CARD_NAME=${CREATED_ASOUND_CARD_NAME:-}
     - ENABLE_SOFTVOLUME=${ENABLE_SOFTVOLUME:-yes}
     - ENABLE_GENERATED_TONE=${ENABLE_GENERATED_TONE:-yes}
     - ASOUND_FILE_PREFIX=${ASOUND_FILE_PREFIX:-}
     - FRIENDLY_NAME=${FRIENDLY_NAME:-TidalConnect}
     - MODEL_NAME=${MODEL_NAME:-Audio Streamer}
     - MQA_CODEC=${MQA_CODEC:-false}
     - MQA_PASSTHROUGH=${MQA_PASSTHROUGH:-false}
     - FORCE_PLAYBACK_DEVICE=${FORCE_PLAYBACK_DEVICE:-}
     - SLEEP_TIME_SEC=${SLEEP_TIME:-3}
     - RESTART_ON_FAIL=${RESTART_ON_FAIL:-1}
     - RESTART_WAIT_SEC=${RESTART_WAIT_SEC:-10}
     - LOG_LEVEL=${LOG_LEVEL:-3}
     - DISABLE_CONTROL_APP=${DISABLE_CONTROL_APP:-0}
     - DISABLE_APP_SECURITY=${DISABLE_APP_SECURITY:-}
     - DISABLE_WEB_SECURITY=${DISABLE_WEB_SECURITY:-}
    volumes:
     - ./assets:/assets:ro
     - ./userconfig:/userconfig:ro
     - ./bin/common.sh:/common.sh
     - ./bin/entrypoint.sh:/entrypoint.sh
     - /var/run/dbus:/var/run/dbus
    dns:
      - ${DNS_SERVER_LIST:-8.8.8.8}
    restart: unless-stopped
pi@moode-living-digi-desktop:~/git/tidal-connect $ cat .env
FRIENDLY_NAME=Moode Living Digi Desktop
CARD_NAME=sndrpihifiberry
```

**Reply by ponim** [reply]
thanks! mine is 100% the same:

FRIENDLY_NAME=Moode DAC
MODEL_NAME=Raspberry Pi 4b
CARD_NAME=sndrpihifiberry

version: "3"

services:
  tidal-connect:
    image: ${TIDAL_CONNECT_IMAGE:-edgecrush3r/tidal-connect:latest}
    container_name: tidal-connect
    network_mode: host
    devices:
     - /dev/snd
    environment:
     - CARD_NAME=${CARD_NAME:-}
     - CARD_INDEX=${CARD_INDEX:-}
     - CARD_DEVICE=${CARD_DEVICE:-}
     - CARD_FORMAT=${CARD_FORMAT:-}
     - CERTIFICATE_PATH=${CERTIFICATE_PATH:-}
     - CLIENT_ID=${CLIENT_ID:-}
     - CREATED_ASOUND_CARD_NAME=${CREATED_ASOUND_CARD_NAME:-}
     - ENABLE_SOFTVOLUME=${ENABLE_SOFTVOLUME:-no}
     - ENABLE_GENERATED_TONE=${ENABLE_GENERATED_TONE:-yes}
     - ASOUND_FILE_PREFIX=${ASOUND_FILE_PREFIX:-}
     - FRIENDLY_NAME=${FRIENDLY_NAME:-TidalConnect}
     - MODEL_NAME=${MODEL_NAME:-Audio Streamer}
     - MQA_CODEC=${MQA_CODEC:-false}
     - MQA_PASSTHROUGH=${MQA_PASSTHROUGH:-false}
     - FORCE_PLAYBACK_DEVICE=${FORCE_PLAYBACK_DEVICE:-}
     - SLEEP_TIME_SEC=${SLEEP_TIME:-3}
     - RESTART_ON_FAIL=${RESTART_ON_FAIL:-1}
     - RESTART_WAIT_SEC=${RESTART_WAIT_SEC:-10}
     - LOG_LEVEL=${LOG_LEVEL:-3}
     - DISABLE_CONTROL_APP=${DISABLE_CONTROL_APP:-0}
     - DISABLE_APP_SECURITY=${DISABLE_APP_SECURITY:-}
     - DISABLE_WEB_SECURITY=${DISABLE_WEB_SECURITY:-}
    volumes:
     - ./assets:/assets:ro
     - ./userconfig:/userconfig:ro
     - ./bin/common.sh:/common.sh
     - ./bin/entrypoint.sh:/entrypoint.sh
     - /var/run/dbus:/var/run/dbus
    dns:
      - ${DNS_SERVER_LIST:-8.8.8.8}
    restart: unless-stopped

**Reply by GioF71** [reply]
maybe try setting LOG_LEVEL to 4 (or up to 6) in your .env file and see what happens, hopefully we can understand why the app is falling back to a lossy stream.
Also maybe check your network speed with [fast](https://fast.com/) or [cloudflare](https://speed.cloudflare.com/).
Are you using DNS from your ISP? despite the DNS_SERVER_LIST setting, the setting might be overridden by your router. Just saying, of course I am not sure.

**Reply by GioF71** [reply]
post deleted (reveals tokens). But I downloaded it.

**Reply by ponim** [reply]
and the speed is ~350Mbps over WiFi

**Reply by GioF71** [reply]
from what I see, it's receiving a flac stream. Just search for flac on your log

**Reply by ponim** [reply]
oh, right! interesting why I do see the different GUI indication from what you have.

But thanks in any case for your time - important to know that the sound quality I get is OK!!!

**Reply by GioF71** [reply]
> oh, right! interesting why I do see the different GUI indication from what you have.
> 
> But thanks in any case for your time - important to know that the sound quality I get is OK!!!

I might be due to an application glitch... I used the android app, you are using the windows app from what I understand from the fact that it displays chromecast devices and not airplay. But the top-right button looks like some MacOS version.
Edit: it says MacBook Pro... I did not know that the mac version could play to chromecast devices.
Can you try some mobile version? Be sure to disconnect the app from the app you are currently using before trying to connect the mobile version.

**Reply by ponim** [reply]
attached is from my iPhone :)
![PHOTO-2026-03-31-12-53-41](https://github.com/user-attachments/assets/88e38c05-76f9-4a9d-b303-efc7fd19a647)

**Reply by GioF71** [reply]
No indication of stream being lossless, but no indication of it being lossy as well. So we get the silence treatment :-)
It that's the trend, the android app will start to behave similarly sooner than later I think.
Anyway you can always check using the logs.

Remember, this thing might stop working at any time. Consider the alternatives. As you are running apple devices, the easiest is AirPlay. Maybe version 1 (using shairport sync with docker if you don't want to mess with moode's configuration files and potentially brick or even partially disrupt your device functionality) can still stream lossless 16bit/44.1 kHz.

Consider starring the repo if you like the work!

**Reply by ponim** [reply]
thanks again! for me Tidal connect sound quality is unbeatable! AirPlay is always the option, but sounds not hi-end at all!

**Comment by GioF71** [comment]
@mike-golant can you post your .env file? If you set 

```text
ENABLE_SOFTVOLUME=False
```

that should be `yes` or `no`. Try removing the line.
Also post the output of 

```text
cat /proc/asound/cards
```

and 

```text
aplay -L
```

**Comment by Iruizcast** [comment]
Hello!

First of all, thank you for your work. I have Moode Audio (10.1.2 2026-03-02) installed on a Raspberry Pi Zero 2W. After trying several configurations for the userconfig, this is the ASound configuration that worked for me to get sound over HDMI output:

file : `hdmi-rpi-zero.asound.conf `
```
pcm.tidalout {
    type iec958
    slave.pcm "hw:CARD=vc4hdmi,DEV=0"
    slave.format "IEC958_SUBFRAME_LE"
}

```

And this is my .env file
```
FRIENDLY_NAME=RaspberryPi
MODEL_NAME=Raspberry Pi Zero 2
ENABLE_SOFTVOLUME=no
ENABLE_GENERATED_TONE=no
FORCE_PLAYBACK_DEVICE=tidalout
ASOUND_FILE_PREFIX=hdmi-rpi-zero
```
Thank you!!

**Reply by GioF71** [reply]
Awesome! Do you Want to contribute your file(s) and maybe instructions with a pull request?

**Reply by adamnight0-pixel** [reply]
Hi, I wanted to try out Moode Audio with Tidal. Until now, I’ve been using piCorePlayer with its built-in Tidal plugin, but unfortunately, the quality offered by that plugin is limited to 16/44. I’ve installed Moode and am now trying to set up Docker so that Tidal Connect works. Docker is installed, but I’ve reached the step where I need to select the sound card drivers, and I don’t know how to switch to the HiFiBerry Digi+ Pro. I’m using the IanCanada Digi Transport HAT for the Raspberry Pi, which requires those specific drivers. 
Could someone help me out?
 I’m not very tech-savvy with this sort of thing—thanks!

**Reply by GioF71** [reply]
hello, if you're looking for more than 16/44, you're out of luck with this (now old) tidal connect implementation. It plays up to 16/44 in fact. Do you strictly need tidal-connect? Because you can get hires tidal with multiple different alternatives, look at the README file.

**Reply by adamnight0-pixel** [reply]
I don't need Tidal Connect. As I mentioned, I currently play Tidal via picorePlayer and its built-in Tidal plugin (using the Squeezelite application), but I want better sound quality. I also want it to work the same way as before—meaning the phone with the app acts as a remote control, not as the audio streaming source.
I thought Moode with Tidal Connect would give me the best quality; what do you suggest instead? Unfortunately, I don't know anything about programming.

**Reply by GioF71** [reply]
it's this specific old binary that doesn't allow more than 16/44... modern devices like e.g. a WiiM Pro can play 24/192 over tidal connect without an issue.
If you want to use piCorePlayer (so squeezelite), you can consider running Music Assistant, the tidal plugin there works up to 24/192 as well. And squeezelite IMO is the best choice as a playback solution on Music Assistant.

**Reply by adamnight0-pixel** [reply]
Okay, thanks for the explanation. What exactly should I look for? I've been using picoreplayer for years.

**Reply by GioF71** [reply]
have a look at https://www.music-assistant.io/. You can install it using docker. Or on Home Assistant if you have one running. Then you can connect your piCorePlayer to that server instead of lyrion. Only issue is should you want to run it in the same server where lyrion runs. You might need to change some ports for the slimproto protocol.

**Reply by GioF71** [reply]
Or of course you can run a upmpdcli media server with the tidal plugin enabled and use Moode as a renderer. Dislaimer: I am the author fo the plugin (not upmpdcli, only the plugin). See [upmpdcli repo](https://github.com/GioF71/upmpdcli-docker) and [example configuration](https://github.com/GioF71/audio-tools/tree/main/media-servers/tidal-hires).

**Reply by adamnight0-pixel** [reply]
Thanks for the reply. Unfortunately, I can't handle this, so I'm dropping the idea. 
I'll keep listening via the picoreplayer plugin, or eventually just buy a WiiM and sell off all these "magic" add-on boards and power supplies. ;)

**Reply by GioF71** [reply]
> Thanks for the reply. Unfortunately, I can't handle this, so I'm dropping the idea. I'll keep listening via the picoreplayer plugin, or eventually just buy a WiiM and sell off all these "magic" add-on boards and power supplies. ;)

alright, cool. If you change your mind, feel free to drop another comment here.

**Comment by blau-git** [comment]
I have tidal-connect running on moode and would like to output via the camilla DSP pipeline. Camillladsp does not show up as an alsa soundcard but aplay -L finds it. Any suggestions on how to make this work?
Thank you in advance

**Reply by GioF71** [reply]
> I have tidal-connect running on moode and would like to output via the camilla DSP pipeline. Camillladsp does not show up as an alsa soundcard but aplay -L finds it. Any suggestions on how to make this work? Thank you in advance

hello, does the device appear when you do `aplay -L` on the host, or on the container? I believe camilladsp should be installed in the container.
Other option would be to create a loop device and let the container (the tidal-connect application) play on that, and let camilladsp run on the host. I tried that and it worked, but I could not find a way to reliably release the device when nothing is playing or even when the player is disconnected from the Tidal App.
It was only kind of working but would cut the first samples unfortunately. Someone more knowledgeable than me can succeed of course.
Accepting exclusive access at all times would work, but that's a very hard requirement, as you cannot use any other player under this condition (not even Moode itself).
So I did not go further. 
I still have access to the Gemini chat I used to try this if you are interested.

Edit: typos

**Reply by blau-git** [reply]
I did aplay -L on the host. Letting tidal-connect play on a loop device looks sensible to me - how exactly did you do that?
I also observed the problem of not releasing the device even when tidal-connect had played via the hardware device, so that doesn't appear to depend on whether one uses a hardware or a loop device.

**Reply by GioF71** [reply]
Hello, as you're running Moode, you probably already have one or more loopback devices. Can you post the output of aplay -L?
Also, is CamillaDSP already running?
If CamillaDSP is already running, you might get away and achieve your goal by *simply* instructing tidal connect to play on the same output used by mpd. See the `/etc/mpd.conf` and look for the alsa output. Maybe post a few details and I might help you get through.

About the chat I mentioned, I am starting to retrieve some details of the camilladsp aspects, but unfortunately I did not create a dedicated chat for this topic so all the information is buried in a very long chat with Gemini. Which is of course also slow to load. In any case, I was working on a vanilla system, not on moode, so I could freely make changes to alsa and camillaDSP settings. I wouldn't do the same on a moode instance without risking to break something.

**Reply by blau-git** [reply]
Thank you for your quick reply!

Camilladsp is running, and all other audio (stored audio files, radio streams) is rendered through it.

-> output of aplay -L:
null
    Discard all samples (playback) or generate zero samples (capture)
iec958
    IEC958 (S/PDIF) Digital Audio Output
bluealsa
    Bluetooth Audio
_audioout
_audioout__
alsaequal
plug_alsaequal
btstream
camilladsp
crossfeed
plug_bs2b
eqfa12p
plug_eqfa12p
invpolarity
trx_send
hw:CARD=sndrpihifiberry,DEV=0
    snd_rpi_hifiberry_digi, HiFiBerry Digi+ Pro HiFi wm8804-spdif-0
    Direct hardware device without any conversions
plughw:CARD=sndrpihifiberry,DEV=0
    snd_rpi_hifiberry_digi, HiFiBerry Digi+ Pro HiFi wm8804-spdif-0
    Hardware device with all software conversions
default:CARD=sndrpihifiberry
    snd_rpi_hifiberry_digi, HiFiBerry Digi+ Pro HiFi wm8804-spdif-0
    Default Audio Device
sysdefault:CARD=sndrpihifiberry
    snd_rpi_hifiberry_digi, HiFiBerry Digi+ Pro HiFi wm8804-spdif-0
    Default Audio Device
iec958:CARD=sndrpihifiberry,DEV=0
    snd_rpi_hifiberry_digi, HiFiBerry Digi+ Pro HiFi wm8804-spdif-0
    IEC958 (S/PDIF) Digital Audio Output
dmix:CARD=sndrpihifiberry,DEV=0
    snd_rpi_hifiberry_digi, HiFiBerry Digi+ Pro HiFi wm8804-spdif-0
    Direct sample mixing device

-> /etc/mpd.conf
music_directory "/var/lib/mpd/music"
playlist_directory "/var/lib/mpd/playlists"
db_file "/var/lib/mpd/database"
log_file "/var/log/mpd/log"
pid_file "/var/run/mpd/pid"
state_file "/var/lib/mpd/state"
sticker_file "/var/lib/mpd/sticker.sql"
user "mpd"
group "audio"
bind_to_address "any"
port "6600"
log_level "default"
restore_paused "yes"
auto_update "no"
follow_outside_symlinks "yes"
follow_inside_symlinks "yes"
zeroconf_enabled "no"
zeroconf_name "Raspberrypi MPD"
filesystem_charset "UTF-8"
metadata_to_use "+comment"
replaygain "off"
replaygain_preamp "0"
volume_normalization "no"
audio_buffer_size "4096"
max_output_buffer_size "131072"
max_playlist_length "16384"
max_connections "128"

decoder {
plugin "ffmpeg"
enabled "yes"
}

input {
plugin "curl"
proxy ""
proxy_user ""
proxy_password ""
}

resampler {
plugin "soxr"
quality "high"
threads "1"
}

audio_output {
type "alsa"
name "ALSA Default"
device "_audioout"
mixer_type "software"
dop "no"
stop_dsd_silence "no"
thesycon_dsd_workaround "no"
}

audio_output {
type "alsa"
name "ALSA Bluetooth"
device "_audioout"
mixer_type "software"
}

audio_output {
type "httpd"
name "HTTP Server"
port "8000"
encoder "lame"
bitrate "320"
tags "yes"
always_on "yes"
format "44100:16:2"
}

**Comment by GioF71** [comment]
So we need to play to `_audioout`.
I modified my docker-compose.yaml file and added this line under volumes:

```text
    - /etc/alsa/conf.d/_audioout.conf:/etc/asound.conf
```

because there is where _audioout is defined, at least on my box:

```text
pi@moode-living-digi-desktop:~/git/tidal-connect $ cat /etc/alsa/conf.d/_audioout.conf 
#########################################
# This file is managed by moOde
#########################################
pcm._audioout {
type copy
slave.pcm "plughw:1,0"
}
```

Then in my `.env` I removed any CARD_NAME/CARD_INDEX and set it to:

```text
pi@moode-living-digi-desktop:~/git/tidal-connect $ cat .env
FRIENDLY_NAME=Moode Living Digi Desktop
FORCE_PLAYBACK_DEVICE=_audioout
DNS_SERVER_LIST=192.168.1.2
```

I had some DNS errors in the previous run so I specified my pihole host in DNS_SERVER_LIST.
Worked. 
But I don't hava camilladsp configured. Let me know if this is useful.

**Reply by blau-git** [reply]
Nice idea, especially since my /etc/alsa/conf.d/_audioout.conf shows that _audioout is routed via camilladsp:

pcm._audioout {
type copy
slave.pcm "camilladsp"
}

However, when I add FORCE_PLAYBACK_DEVICE=_audioout to my .env tidal-connect disappears from the list of playback devices on my phone client. When I don't, tidal-connect's output is directly rendered to the hardware, bypassing camilladsp, as before.

**Reply by GioF71** [reply]
yes, that's probably because inside the container, camilladsp is not defined. Can you verify where that output `camilladsp`is defined? It should be somewhere in `/etc/alsa/conf.d`.
Maybe we can just build a custom asound.conf. We cannot mount the /etc/alsa/conf.d directory, I tried, the software in the container image is probably too old for this relatively new feature (I don't know when it was introduced).

**Reply by blau-git** [reply]
OK, my /etc/alsa/conf.d/camilladsp.conf:

# For more info about this configuration file see: https://github.com/scripple/alsa>
pcm.camilladsp {
    type cdsp
    cpath "/usr/local/bin/camilladsp"
    config_out "/usr/share/camilladsp/working_config.yml"


    # config_cdsp says to use the new CamillaDSP internal substitutions.
    # When config_cdsp is set to an integer != 0 the hw_params and
    # extra samples are passed to CamillaDSP on the command line as
    # -f format -r samplerate -n channels -e extra_samples
    config_cdsp 1

####################################
# Set the values for your DAC here #
####################################
    min_channels 1
    max_channels 8
    rates = [
        44100
        48000
        88200
        96000
        176400
        192000
        352800
        384000
    ]

###############################################
# Uncomment the following lines,              #
# if you want more detailed logging output in #
# /var/log/mpd/log                            #
###############################################
   cargs [
#        -v
        -p "1234"
        -a "0.0.0.0"
        -s "/var/lib/cdsp/statefile.yml"
   ]

}

**Reply by GioF71** [reply]
uhm, I don't think it will work in the container, because of the dependency with the camilladsp binary.

**Reply by GioF71** [reply]
Hello, I am sorry there is no easy solution. Also with old libraries explicitly downloaded from debian archives to build the tidal connect image, I doubt a precompiled camilladsp binary would work, and compiling would be very slow. And even if we were to be able to have camilladsp installed or built, there would be one instance running on the host and another in the container, and this would probably require to replicate the configuration files somehow.
Let me know what you thing about all of these!
